crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -1230,6 +1234,135 @@ impl ParallelSparseTrie {
         nodes.extend(self.upper_subtrie.nodes.iter());
         nodes
     }
+
+    /// Example implementation of parallel node revealing
+    /// TODO move this onto the trait
+    pub fn reveal_nodes(&mut self, mut nodes: Vec<RevealedSparseNode>) -> SparseTrieResult<()> {
+        if nodes.is_empty() {
+            return Ok(())
+        }
+
+        // Sort nodes first by their subtrie, and secondarily by their path. This allows for
+        // grouping nodes by their subtrie using `chunk_by`.
+        nodes.sort_unstable_by(
+            |RevealedSparseNode { path: path_a, .. }, RevealedSparseNode { path: path_b, .. }| {
+                let subtrie_type_a = SparseSubtrieType::from_path(path_a);
+                let subtrie_type_b = SparseSubtrieType::from_path(path_b);
+                subtrie_type_a.cmp(&subtrie_type_b).then(path_a.cmp(path_b))
+            },
+        );
+
+        // Update the top-level branch node masks. This is simple and can't be done in parallel.
+        for RevealedSparseNode { path, masks, .. } in nodes.iter() {
+            if let Some(tree_mask) = masks.tree_mask {
+                self.branch_node_tree_masks.insert(*path, tree_mask);
+            }
+            if let Some(hash_mask) = masks.hash_mask {
+                self.branch_node_hash_masks.insert(*path, hash_mask);
+            }
+        }
+
+        let mut nodes_slice = nodes.as_slice();
+
+        // Due to the sorting, if there are any nodes to be revealed in the upper trie then the
+        // first node will be an upper trie node. Handle all upper trie reveals using
+        // `ParallelSparseTrie::reveal_node`, which deals with revealing recursively into
+        // lower-subtries as necessary.
+        if SparseSubtrieType::from_path(&nodes[0].path).lower_index().is_none() {
+            let mut num_upper_nodes = 0;
+            let upper_nodes =
+                nodes.iter().take_while(|n| SparseSubtrieType::path_len_is_upper(n.path.len()));
+
+            for node in upper_nodes {
+                // TODO once reveal_node is not on the trait we can probably make it accept
+                // references, and skip these clones.
+                self.reveal_node(node.path, node.node.clone(), node.masks.clone())?;
+                num_upper_nodes += 1;
+            }
+
+            // truncate the upper trie nodes off the front of the nodes slice.
+            nodes_slice = &nodes_slice[num_upper_nodes..];
+        }
+
+        #[cfg(not(feature = "std"))]
+        // Reveal lower subtrie nodes serially if nostd
+        for node in nodes_slice {
+            for node in nodes {
+                self.reveal_node(node.path, node.node.clone(), node.masks.clone())?;
+            }
+        }
```

Comment:
why the double loop here? nodes should also still be owned here so I think we can just do `for node in nodes`?